# <u>**Extract, Transform and Load (ETL)**</u>

This notebook is the intial step toward a standard data analytics workflow. Data quality is key to ensure effective Exploratory Data Analysis, Feature Engineering, Machine Learning Model development and dashboard visualisation. 

The data cleaning process requires thorough checks to ensure the integrity of the datasets are reviewed before saving transformations into a joined DataFrame. The steps taken in this notebook are:

- Load libraries and datasets

- Initial datasets review 

- Standardise column names 

- Identify missing, unique or duplicate values

- Ammend DataTypes 

- Add year columns for year-on-year analysis

- Join 2023-2024 and 2025-2025 datasets

- Drop identifiable information columns  

- Save DataFrame into clean data folder

ETL can reveal additional demands from the datasets during these checks. This plan is based on the most common steps taught by Code Institute LMS content, tutors and online resources included in the credits section of the README. Built-in Visual Studio Code Copilot will also collaboratively fix, modify and add comments to streamline the process.

</b>

## <u>**Load Libraries and Data**</u>

Libraries - Pandas and Numpy - are imported. Raw datasets exported as CSV from EXCEL sheets are loaded into two seperate DataFrames using Pandas.  

In [1]:
# Import libraries 
import pandas as pd
import numpy as np

# Read the CSV file into a DataFrame
df_24 = pd.read_csv('/Users/admin/Documents/Code Institute/ace_project/data/raw/2023_24_ace_raw.csv')
df_26 = pd.read_csv('/Users/admin/Documents/Code Institute/ace_project/data/raw/2025_26_ace_raw.csv')

# Check rows and columns in the DataFrames
print("Shape of df_24:", df_24.shape)
print("Shape of df_26:", df_26.shape)

Shape of df_24: (3085, 11)
Shape of df_26: (2369, 11)


</b>

### <u>**Intial DataFrame Review**</u>

Use <code>.info</code> and <code>.head()</code> for an intial review on both DataFrames.

In [2]:
df_24.info()  # Display information about df_24
df_24.head()  # Display the first few rows of df_24

<class 'pandas.DataFrame'>
RangeIndex: 3085 entries, 0 to 3084
Data columns (total 11 columns):
 #   Column                 Non-Null Count  Dtype
---  ------                 --------------  -----
 0   Recipient              3085 non-null   str  
 1   Activity name          3085 non-null   str  
 2   Award amount           3085 non-null   str  
 3   Award date             3085 non-null   str  
 4   Decision Month         3085 non-null   str  
 5   Decision quarter       3085 non-null   str  
 6   ACE Area               3085 non-null   str  
 7   Local authority        3085 non-null   str  
 8   Main discipline        3085 non-null   str  
 9   Strand                 3085 non-null   str  
 10  Time-Limited Priority  240 non-null    str  
dtypes: str(11)
memory usage: 265.2 KB


,Recipient,Activity name,Award amount,Award date,Decision Month,Decision quarter,ACE Area,Local authority,Main discipline,Strand,Time-Limited Priority
0,Feelprime Ltd,Old Fire Station Digital,"£27,689",03-Apr-23,April,Q1,South West,"Bournemouth, Christchurch and Poole",Music,"£30,000 and below",Supporting Grassroots Live Music
1,Downtown Pompey,Sticky Floors and Palm Trees,"£29,623",03-Apr-23,April,Q1,South West,Portsmouth,Theatre,"£30,000 and below",NaN
2,Chew Valley Performing Arts,Autumn Fringe Festival 2023,"£25,256",03-Apr-23,April,Q1,South West,Bath and North East Somerset,Combined arts,"£30,000 and below",NaN
3,Upfest,Weston Wallz 2023,"£30,500",03-Apr-23,April,Q1,South West,"Bristol, City of",Visual arts,"£30,000 and below",NaN
4,Chloe Thurlow,"Haunt, Pray, Love","£20,060",03-Apr-23,April,Q1,South West,South Gloucestershire,Theatre,"£30,000 and below",NaN


In [3]:
df_26.info()  # Display information about df_26
df_26.head()  # Display the first few rows of df_26

<class 'pandas.DataFrame'>
RangeIndex: 2369 entries, 0 to 2368
Data columns (total 11 columns):
 #   Column                 Non-Null Count  Dtype
---  ------                 --------------  -----
 0   Recipient              2369 non-null   str  
 1   Activity name          2369 non-null   str  
 2   Award amount           2369 non-null   str  
 3   Award date             2369 non-null   str  
 4   Decision Month         2369 non-null   str  
 5   Decision quarter       2369 non-null   str  
 6   ACE Area               2369 non-null   str  
 7   Local authority        2368 non-null   str  
 8   Main discipline        2369 non-null   str  
 9   Strand                 2369 non-null   str  
 10  Time-Limited Priority  224 non-null    str  
dtypes: str(11)
memory usage: 203.7 KB


,Recipient,Activity name,Award amount,Award date,Decision Month,Decision quarter,ACE Area,Local authority,Main discipline,Strand,Time-Limited Priority
0,Window Wanderland CIC,"Community of practise, Wanderland Pilot","£33,522",01-Apr-25,April,Q1,South West,"Bristol, City of",Combined arts,"£30,000 and below",NaN
1,Cat and Mouse Theatre,Time Crime,"£27,461",01-Apr-25,April,Q1,South West,South Gloucestershire,Theatre,"£30,000 and below",NaN
2,London Borough of Ealing Libraries and communi...,Y-Our Story Visual Stories of Acton,"£27,000",01-Apr-25,April,Q1,London,Ealing,Libraries,"£30,000 and below",Universal Library Offers
3,Exmouth Festival,Exmouth Festival 2025,"£38,500",01-Apr-25,April,Q1,South West,East Devon,Combined arts,"Above £30,000",NaN
4,Crediton Heart Project,Crediton Heart Summer Arts Festival,"£15,135",01-Apr-25,April,Q1,South West,Mid Devon,Combined arts,"£30,000 and below",NaN


The outputs for <code>.info()</code> shows both DataFrames share the same titles for column names but will require text standardisation. 

Missing values are present in both DataFrames <code>df_24</code> shows 240 rows contain values in Time-Limited Priority column. Meanwhile, <code>df_26</code> has 224 values for Time-Limited Priority and one missing value in the Local Authority column. 

DataTypes are all set as strings, which require ammending into numerical, data-time and categorical types to plot visualisation in the EDA notebook.

Identifiable information is shown in the Recipient and Activity name from the output for <code>.head()</code>. These columns must be dropped to ensure ethical data handling. 

</b>

## <u>**Standardise Column Names**</u>

Format must be consistent to access columns in the EDA, Feature Engineering and ML Modeling notebooks. Remove leading or trailing whitespace use <code>.strip()</code>, replace spaces or hyphens with underscores use <code>.replace()</code> and convert text to lowercase use <code>.lower()</code>. These functions are a professional standard protocol for technical users to navigate the code cells.  

In [4]:
# Format column names
df_24.columns = (df_24.columns
                 .str.strip()
                 .str.lower()
                 .str.replace(' ', '_'))

df_26.columns = (df_26.columns
                 .str.strip()
                 .str.lower()
                 .str.replace(' ', '_'))


In [5]:
# Check column names 2023-2024
df_24.head() 

,recipient,activity_name,award_amount,award_date,decision_month,decision_quarter,ace_area,local_authority,main_discipline,strand,time-limited_priority
0,Feelprime Ltd,Old Fire Station Digital,"£27,689",03-Apr-23,April,Q1,South West,"Bournemouth, Christchurch and Poole",Music,"£30,000 and below",Supporting Grassroots Live Music
1,Downtown Pompey,Sticky Floors and Palm Trees,"£29,623",03-Apr-23,April,Q1,South West,Portsmouth,Theatre,"£30,000 and below",NaN
2,Chew Valley Performing Arts,Autumn Fringe Festival 2023,"£25,256",03-Apr-23,April,Q1,South West,Bath and North East Somerset,Combined arts,"£30,000 and below",NaN
3,Upfest,Weston Wallz 2023,"£30,500",03-Apr-23,April,Q1,South West,"Bristol, City of",Visual arts,"£30,000 and below",NaN
4,Chloe Thurlow,"Haunt, Pray, Love","£20,060",03-Apr-23,April,Q1,South West,South Gloucestershire,Theatre,"£30,000 and below",NaN


In [6]:
# Check column names 2025-2026
df_24.head()

,recipient,activity_name,award_amount,award_date,decision_month,decision_quarter,ace_area,local_authority,main_discipline,strand,time-limited_priority
0,Feelprime Ltd,Old Fire Station Digital,"£27,689",03-Apr-23,April,Q1,South West,"Bournemouth, Christchurch and Poole",Music,"£30,000 and below",Supporting Grassroots Live Music
1,Downtown Pompey,Sticky Floors and Palm Trees,"£29,623",03-Apr-23,April,Q1,South West,Portsmouth,Theatre,"£30,000 and below",NaN
2,Chew Valley Performing Arts,Autumn Fringe Festival 2023,"£25,256",03-Apr-23,April,Q1,South West,Bath and North East Somerset,Combined arts,"£30,000 and below",NaN
3,Upfest,Weston Wallz 2023,"£30,500",03-Apr-23,April,Q1,South West,"Bristol, City of",Visual arts,"£30,000 and below",NaN
4,Chloe Thurlow,"Haunt, Pray, Love","£20,060",03-Apr-23,April,Q1,South West,South Gloucestershire,Theatre,"£30,000 and below",NaN


</b>

## <u>**Missing Values**</u>

### **Inspect Missing Values**

Missing values can lower the accuracy of visualisations in the EDA and ML model performance. NaN or blank values can also cause errors for tools and libraries in the ML Modeling notebook. To inspect missing values use <code>.isna().sum()</code>.

In [7]:
# Check for missing values in df_24
df_24.isna().sum()

recipient                   0
activity_name               0
award_amount                0
award_date                  0
decision_month              0
decision_quarter            0
ace_area                    0
local_authority             0
main_discipline             0
strand                      0
time-limited_priority    2845
dtype: int64

Time-Limited priority has 2,854 missing values for awards received between 2023-2024. Further research to establish the purpose of this column can help determine how to handle these missing values.

In [8]:
# Check for missing values in df_26
df_26.isna().sum()  

recipient                   0
activity_name               0
award_amount                0
award_date                  0
decision_month              0
decision_quarter            0
ace_area                    0
local_authority             1
main_discipline             0
strand                      0
time-limited_priority    2145
dtype: int64

Again, the time-limited priority column flags 2,145 missing values that need to be considered prior to transformation. This DataFrame for awards received between 2025-2026 also contains one missing value in the Local Authorities column.

In [9]:
df_26[df_26["local_authority"].isna()]

,recipient,activity_name,award_amount,award_date,decision_month,decision_quarter,ace_area,local_authority,main_discipline,strand,time-limited_priority
1328,Comics Cultural Impact Collective,NLPG O30 Oct25,"£37,000",17-Dec-25,December,Q3,North,NaN,Literature,"Above £30,000",NaN


According to official Arts Council documentation:

>We define place as a specific geographic area of any scale (village, town, local authority, city or city region) where the proposed intervention will take place and where a cultural strategy (or similar) exists that covers the same geography. This could mean that an application might be for activity in a town or village that is covered by a cultural strategy for a wider district.

[National Lottery Project Grants Information Sheet](https://www.artscouncil.org.uk/sites/default/files/2023-09/Place%20Partnership%20projects%20and%20Project%20Grants%20-%20Information%20sheet.pdf)

The Arts Council website also addresses the Time-limited priority column. The missing values represent projects that are not time-limited. 

[National Lottery Project Grants: Guidance Library](https://www.artscouncil.org.uk/ProjectGrants/national-lottery-project-grants-guidance-library#t-in-page-nav-5)

</b>

### **Handle Missing Values**

Replace missing value in the 2025-2026 DataFrame for <code>local_authority</code> with "Unknown", as this recipient potentially operates in more than one local authority according to the Arts Council documentation. 

In [ ]:
# Replace missing value in the 'local_authority' column with 'Unknown'
df_26['local_authority'] = df_26['local_authority'].fillna('Unknown')

# Check for remaining missing values in df_26
df_26.isna().sum()

recipient                   0
activity_name               0
award_amount                0
award_date                  0
decision_month              0
decision_quarter            0
ace_area                    0
local_authority             0
main_discipline             0
strand                      0
time-limited_priority    2145
dtype: int64

## **<u>Convert DataTypes</u>**

Data consistency is neccessary to plot visualisations in the EDA notebook. Review DataTypes and convert each column to numerical, date-time or categorical. Use </code><code>.dtypes()</code> to display the current DataTypes in each DateFrame. 

The DataFrames contain mostly categorical columns, one numeric value and date-time column which were identified as strings in the intial review of both DataFrames.

In [ ]:
# Check DataTypes for 2023-2024 and 2025-2026
print("Data types for df_24:")
print(df_24.dtypes)

print("\nData types for df_26:")
print(df_26.dtypes)

</b >

### **Award Amount**

Similar to standardising column names in the previous section <code>.replace()</code> and <code>.strip()</code> remove symbols and hidden spaces to avoid ValueError caused by an invalid literal. Convert string into numeric DataType use <code>.astype(int)</code>.

In [ ]:
# Clean and convert the 'award_amount' column in df_24
df_24["award_amount"] = (df_24["award_amount"]
                         .str.replace(',','')
                         .str.replace('£','')
                         .str.strip()
                         .astype(int))


# Clean and convert the 'award_amount' column in df_26
df_26["award_amount"] = (df_26["award_amount"]
                         .str.replace(',','')
                         .str.replace('£','')
                         .str.strip()
                         .astype(int))


# Check the data type of the 'award_amount' column
print("Data type df_24:", df_24["award_amount"].dtype) 
print("Data type df_26:", df_26["award_amount"].dtype) 


</b>

### **Date-Time**

Convert award-date column from string into date-time. Use <code>pd.to_datetime</code> for both DateFrames and check the conversion is successful.  

In [ ]:
# Convert award_date columns to datetime so .dt.year works
df_24['award_date'] = pd.to_datetime(df_24['award_date'], errors='coerce')
df_26['award_date'] = pd.to_datetime(df_26['award_date'], errors='coerce')
df_24['award_year'] = df_24['award_date'].dt.year

# Extract the year from the award_date column in df_26
df_26['award_year'] = df_26['award_date'].dt.year

In [ ]:
# Check the first few rows of df_24 to verify the changes
df_24.head()   

Intially, the built-in copilot modified the code to remove time-stamp in the column with <code>.normalize()</code> which did not solve the issue because its function is to reset the time to midnight. Instead, <code>.dt.date</code> was the appropiate method to remove the time-stamp from column.

However, this causes an obstacle to extract the year to create a column in the next step. Thus, prior to removing the time-stamp add column then remove time-stamp.

Add a column dedicated to year and improve filtering or grouping rows. Use <code>.dt.year</code> to extract the year from award date column.  

In [ ]:
# Remove time from award_date column in df_24 and df_26
df_24['award_date'] = df_24['award_date'].dt.date
df_26['award_date'] = df_26['award_date'].dt.date

In [ ]:
# Check the first few rows of df_24 to verify the changes
df_24.head()  

In [ ]:
# Check the first few rows of df_26 to verify the changes
df_26.head()

</b>

### **Categorical**

Categorical data types contain a list of unique values that represent different categories. Use <code>.astype("category")</code> to convert strings into a categorical data type. 

In [ ]:
# Group categorical columns for conversion to 'category' data type
categorical_cols = [
    "main_discipline",
    "strand",
    "time_limited_priority",
    "ace_area"
]

# Convert the categorical columns
for df in [df_24, df_26]:
    for col in categorical_cols:
        if col in df.columns:
            df[col] = df[col].astype("category")


In [ ]:
# check the data types of the categorical columns in df_24
df_24.dtypes

In [ ]:
# check the data types of the categorical columns in df_24
df_24.dtypes

# check the data types of the categorical columns in df_26
df_26.dtypes

Time-limited priority column remained as a string. In the intial review missing values were identified which will impact coversion. Thus, missing, duplicate and unique values must be assessed prior to converting data types. 